### ISSUE TITLE
Keep others’ IPNS records alive

### ISSUE THREAD

**Issue by ion1** [ipfs/go-ipfs#1958]
```
ipfs name keep-alive add <friend’s node id>
```

Periodically get and store the IPNS record and keep serving the latest seen version to the network until the record’s EOL.

**Comment by lgierth** [comment]
You'll be able to pin IPNS records like anything else once we have IPRS

**Comment by ion1** [comment]
Awesome

**Comment by koalalorenzo** [comment]
Waiting for this feature 👍

**Comment by Falsen** [comment]
But doesn't it make more sense if they are automatically pinned by nodes? Or would it be resource heavy,?

**Comment by koalalorenzo** [comment]
Consider that if pinned those have to be updated constantly via signatures etc etc...

**Comment by Stebalien** [comment]
The issue here is that the signature on IPNS records currently *expire*. We expire them because the DHT isn't persistent and will eventually forget these records anyways. When it does, an attacker would be able to replay an old IPNS record from any point in time.

**Comment by lockedshadow** [comment]
> When it does, an attacker would be able to replay an old IPNS record from any point in time.

Is it really considered more dangerous than possibility of practically disappearing whole materials published under certain IPNS key if **one** (just one!) publisher node with its private key once disappears too? Doesn't this publisher node look like the central point of failure? Outdated, but _valid_ records are really worse than **no records at all**?

I think that ability to replay is not an critical security issue, at least in condition that user is explicitly notified that the obtained result could be outdated. After all, «it will always return _valid_ records (even if a bit stale)», as mentioned in [0.4.18 changelog](https://github.com/ipfs/go-ipfs/releases/tag/v0.4.18).

So what do you think about `--show-publish-time` flag on `ipfs name resolve` command? Is IPNS records itself contain those information?

**Comment by Stebalien** [comment]
@lockedshadow I've been thinking about (and discussing this) this and, well, you're right. Record authors should be able to specify a timeout but there's no reason to remove expired records from the network. Whether or not to _accept_ an expired record would be up to the client.

**Comment by T0admomo** [comment]
@Stebalien What is the best way to go about introducing this change to the protocol?

**Comment by aschmahmann** [comment]
@T0admomo since this is a client and UX change rather than a spec one mostly I would propose what the UX should be along with the various changes that would need to happen in order to enable it.

Some related issues: #7572 #4435 #3117

**Comment by 2color** [comment]
> The issue here is that the signature on IPNS records currently expires and random nodes won't be able to re-sign them as they'd need the associated private key.

According to [the IPNS spec](https://github.com/ipfs/specs/blob/main/IPNS.md#ipns-record), the signature contains the concatenated `value`, `validity`, and `validityType` fields. 

If I understand it correctly, that means that as long as **`validity` is in the future**, there's no reason why nodes wouldn't republish the IPNS record.

Moreover, since that's controlled by the key holder, they have the flexibility to pick any validity at the potential cost of users getting an expired/stale record. This is arguably better than getting no resolution as [pointed out](https://github.com/ipfs/kubo/issues/1958#issuecomment-444201606) by @lockedshadow 

Am I understanding this correctly?

**Comment by bertrandfalguiere** [comment]
> That means that as long as **`validity` is in the future**, there's no reason why nodes wouldn't republish the IPNS record.
> 

I think this could be an attack vector as a malicious node could publish a lot of signed records with infinite validity. They will accumulate on the DHT and clog it sooner or later, and never be flushed out. 

So other clients needs to reject very old records, even if the original publisher wanted them to have very long validity.

**Comment by 2color** [comment]
> I think this could be an attack vector as a malicious node could publish a lot of signed records with infinite validity. They will accumulate on the DHT and clog it sooner or later, and never be flushed out.

I recently read that DHT nodes will drop stored values after ~24 hours, no matter what Lifetime and TTL you set. So it's not really possible to clog the DHT or use this as an attack vector. 

As far as I understand, clients don't reject old records as they have no way of knowing a record's age, they just drop them after 24 hours, when a newer sequence comes or once they expire (the earliest of the three). 

> (An attacker could also spawn many nodes and publish records from them, with the same effect)

I believe that this is what [Fierro](https://github.com/Fierro-Labs/Fierro) allows you to do, though without any malicious intent.

**Comment by bertrandfalguiere** [comment]
> As far as I understand, clients don't reject old records as they have no way of knowing a record's age, they just drop them after 24 hours, when a newer sequence comes or once they expire (the earliest of the three).
> 

Yes, you're right. Droping records is not based on age, I oversimplified. The point is that they are not in the DHT after some time if they are not republished, so they can't accumulate.

**Comment by cornwarecjp** [comment]
When keeping someone else's IPNS record alive, what do you do when you learn about a new record for the same name? I see these possibilities:

* Replace the old record with the new one, and keep the new one alive. On the surface, this seems the most reasonable behavior, since normally in IPNS, you *want* people to learn (only) the new record. I'd call this "pinning the name".
* Keep the old record and ignore the new one. I'd call this "pinning the record". Sounds insane, and I'm not sure whether it's a protocol violation, but I'll give a use case below.
* Keep the old *and* the new, combining the best of both worlds. Disadvantage: anyone requesting an IPNS lookup needs to specify which record they want.

An IPNS record is typically of little use without the data to which it points. I guess, in many applications, someone keeping the IPNS name alive might also want to (recursively) keep the pointed-to data alive ("recursive pinning"). If you've recursively pinned a name, and you receive an update for that name, that'd make you unpin the old pointed-to data, and pin the new pointed-to data. One potential issue with this is that the new data might be arbitrarily large, and therefore much larger than the storage space you'd be willing to spend on it. "Pinning the record" does not have this issue.

There are applications where receiving old data isn't harmful, and where receiving old data is always better than receiving no data. For such applications, "pinning the record" might be the preferred choice, in combination with an application process that gets to decide what to do with a record update. It might, for instance, make an application-level choice to pin only certain parts of the pointed-to DAG, to stay below a storage quotum. And only once the pointed-to data is (partially) downloaded and pinned, the application will replace the old pinned-to record with the new one.

**Comment by cornwarecjp** [comment]
As a poor man's solution, wouldn't it be possible to have an application run besides Kubo, which periodically polls Kubo for the name? If I understand correctly, Kubo caches objects for 24 hours after the last time they were touched, so if the application asks, say, every 12 hours for the name, it'll always stay in Kubo's cache.

As a bonus, the application could store a copy of the latest record received for the name. If Kubo somehow still loses the name, the application can re-upload the last-known record to Kubo[*]. This would double the storage requirement for names, but name records shouldn't be that big.

[*] apparently `/api/v0/routing/put` or `ipfs routing put` can do that, see #10484

**Comment by lidel** [comment]
Multiple issues tracking this, let's continue in https://github.com/ipfs/kubo/issues/4435#issuecomment-3549270256
